octobot/community/feeds/community_mqtt_feed.py

```diff
@@ -0,0 +1,334 @@
+    # The message may be sent in three ways:
+    # QoS 0: the message will be received at most once (also known as “fire and forget”).
+    # QoS 1: the message will be received at least once.
+    # QoS 2: the message will be received exactly once.
+    # from https://www.scaleway.com/en/docs/iot/iot-hub/concepts/#quality-of-service-levels-(qos)
+    default_QOS = 1
+
+    def __init__(self, feed_url, authenticator):
+        super().__init__(feed_url, authenticator)
+        self.mqtt_version = self.MQTT_VERSION
+        self.mqtt_broker_port = self.MQTT_BROKER_PORT
+        self.default_QOS = self.default_QOS
+
+        self._mqtt_client: gmqtt.Client = None
+        self._device_uuid: str = None
+        self._fetching_uuid = False
+        self._subscription_attempts = 0
+        self._fetched_uuid = asyncio.Event()
+        self._subscription_topics = set()
+        self._reconnect_task = None
+        self._processed_messages = set()
+
+    async def start(self):
+        await self.get_or_fetch_device_uuid()
+        await self._connect()
+
+    async def stop(self):
+        self.should_stop = True
+        await self._stop_mqtt_client()
+        if self._reconnect_task is not None and not self._reconnect_task.done():
+            self._reconnect_task.cancel()
+
+    async def _stop_mqtt_client(self):
+        if self._mqtt_client is not None and self._mqtt_client.is_connected:
+            await self._mqtt_client.disconnect()
+
+    async def register_feed_callback(self, channel_type, callback, identifier=None):
+        topic = self._build_topic(channel_type, identifier)
+        try:
+            self.feed_callbacks[topic].append(callback)
+        except KeyError:
+            self.feed_callbacks[topic] = [callback]
+        if topic not in self._subscription_topics:
+            self._subscription_topics.add(topic)
+            self._subscribe((topic, ))
+
+    async def _create_new_device(self):
+        create_query = """
+        mutation CreateDevice($user_id: ObjectId) {
+          insertOneDevice(data: {user_id: $user_id}) {
+            _id
+          }
+        }
+        """
+        await self.authenticator.gql_login_if_required()
+        create_variables = {"user_id": self.authenticator.gql_user_id}
+        error_message = "Error when creating mqtt device"
+        try:
+            create_resp = await self.authenticator.async_graphql_query(create_query, variables=create_variables)
+        except RuntimeError as e:
+            raise RuntimeError(f"{error_message}, {e}")
+        if create_resp.status == 200:
+            json_resp = await create_resp.json()
+            return json_resp["data"]["insertOneDevice"]["_id"]
+        else:
+            raise RuntimeError(f"{error_message}, code: {create_resp.status}, text: {await create_resp.text()}")
+
+    async def _fetch_device_uuid(self, device_id):
+        select_query = """
+        query SelectDeviceUUID($_id: ObjectId, $user_id: ObjectId) {
```

Comment:
Shouldn't we store queries and mutations in a separate module?

Reply:
right